### PR TITLE
Update dataset-list.R

### PR DESCRIPTION
**Pull request by seabbs**
This PR is currently an outline of the work needed to update the UK estimates for test-positive cases and to add new estimates for admissions and deaths. This is in response to an urgent gov request and should be the priority for any/all dev effort. 

@kathsherratt I have updated the list with the skeleton of the items needed and text indicating what goes where. Feel free to ping if not clear. 

@joeHickson in #61 I added the support that this PR draws on so if you could review that ASAP that would be great. In addition, I may have missed something obvious here or there may be problems I haven't seen (such as what will happen to logging)

New delays should go in the `data` folder @kathsherratt.

### CODE REVIEW

**Comment by kathsherratt** [comment]
I've added the new items for UK public data sources (cases, deaths, hospital admissions) from `covidregionaldata`. (NB The call to `covidregionaldata` needs the new `nhsregion = TRUE` option.) 

May be mistakes from unfamiliarity here so happy for others to suggest / edit where needed.

**Review by seabbs — COMMENTED** [review]
This looks good to me. @joeHickson thoughts? Any chance we can get a manual run of this somewhere? I am currently doing SPI-M estimates but may be able to look at that this afternoon.

File: R/dataset-list.R [hunk 1]
```diff
@@ -14,7 +14,7 @@ datasets <- c(
                                                               by = c("date")][, region := "United Kingdom"]
                cases <- data.table::rbindlist(list(cases, uk_cases),
                                                fill = TRUE, use.names = TRUE)},
-             stan_args = list(nhsregions = TRUE)),
+             data_args = list(nhsregions = TRUE)),
   Region$new(name = "united-kingdom",
```
Comment (joeHickson) [inline R/dataset-list.R:18]: names should be unique - use folder_name and dataset_folder_name to locate the folder.

e.g.
name = united-kingdom-admissions, folder_name = united-kingdom, dataset_folder_name = admissions

**Comment by joeHickson** [comment]
I'll give it a spin - I need to get the data sets to have a unique name to target them so I'll tweak those myself.

**Comment by joeHickson** [comment]
No joy - it's erroring. Just merging latest from master over and we'll see if that helps

**Comment by joeHickson** [comment]
new error message now!
```
:/home/covid# Rscript R/run-region-updates.R -w -i "none" --log /home/data/uk_new_log.txt

Attaching package: ‘lubridate’

The following objects are masked from ‘package:base’:

    date, intersect, setdiff, union

Error in gzfile(file, "rb") : cannot open the connection
Calls: source ... eval -> <Anonymous> -> <Anonymous> -> readRDS -> gzfile
In addition: Warning message:
In gzfile(file, "rb") :
  cannot open compressed file '/home/covid/data/cocin-onset-to-death-delay.rds', probable reason 'No such file or directory'
Execution halted
```

**Comment by joeHickson** [comment]
```
2020-10-06 11:28:01 DEBUG skipping location united-kingdom as it is not in the include list
2020-10-06 11:28:01 INFO Processing dataset for united-kingdom-deaths
2020-10-06 11:28:01 INFO Getting regional data
Arranging data by NHS region. 
Also adding new variable: hosp_new_first_admissions. This is NHS data for first hospital admissions, which excludes readmissions. This is available for England and English regions only.
trying URL 'https://www.england.nhs.uk/statistics/wp-content/uploads/sites/2/2020/10/COVID-19-daily-admissions-20201005.xlsx'
Content type 'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet' length 27151 bytes (26 KB)
==================================================
downloaded 26 KB

2020-10-06 11:28:03 TRACE Modifying data
2020-10-06 11:28:03 INFO Processing dataset for united-kingdom-admissions
2020-10-06 11:28:03 INFO Getting regional data
Arranging data by NHS region. 
Also adding new variable: hosp_new_first_admissions. This is NHS data for first hospital admissions, which excludes readmissions. This is available for England and English regions only.
trying URL 'https://www.england.nhs.uk/statistics/wp-content/uploads/sites/2/2020/10/COVID-19-daily-admissions-20201005.xlsx'
Content type 'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet' length 27151 bytes (26 KB)
==================================================
downloaded 26 KB

2020-10-06 11:28:04 TRACE Modifying data
2020-10-06 11:28:04 DEBUG skipping location united-states as it is not in the include list
2020-10-06 11:28:04 DEBUG skipping location regional-cases as it is not in the include list
```

**Comment by joeHickson** [comment]
It seems to by crashing out on `national_deaths <- deaths[region %in% c("England", "Scotland", "Wales", "Northern Ireland"),]` - I can see it completing the deaths <- deaths line before but not beyond

**Comment by seabbs** [comment]
if its a `data.table` it doesn't need the comma - one sec

**Comment by seabbs** [comment]
Think the latest PR should have fixed that issue @joeHickson - as a general point might need a try catch for the user input function as it can be literally anything! I found this https://rdrr.io/cran/futile.logger/man/ftry.html which looks useful.

**Comment by seabbs** [comment]
any thoughts @kathsherratt - can you run the `get_regional_data %>% case_modifier()` outside of the infra?

**Comment by joeHickson** [comment]
re: try / catch - it doesn't crash out any other data sets as that's already isolated.

That's still failing - I can see that it is reaching that line and there are regions present.

```R
futile.logger::flog.debug("get national")
print(unique(deaths$region))
national_deaths <- deaths[region %in% c("England", "Scotland", "Wales", "Northern Ireland")]
futile.logger::flog.debug("get uk wide")
```

2020-10-06 12:51:26 TRACE Modifying data
2020-10-06 12:51:26 DEBUG get national
 [1] "East of England"          "England"                 
 [3] "London"                   "Midlands"                
 [5] "North East and Yorkshire" "North West"              
 [7] "Northern Ireland"         "Scotland"                
 [9] "South East"               "South West"              
[11] "Wales"                   
2020-10-06 12:51:26 INFO Processing dataset for united-kingdom-admissions <- moved straight on

I'm trying to debug on the server though which is not ideal...

**Comment by seabbs** [comment]
This would be so we could see this error message and hence debug the user-supplied function more easily.

**Comment by joeHickson** [comment]
the ftry gave more output:

2020-10-06 12:59:54 TRACE Modifying data
2020-10-06 12:59:55 DEBUG get national
 [1] "East of England"          "England"                 
 [3] "London"                   "Midlands"                
 [5] "North East and Yorkshire" "North West"              
 [7] "Northern Ireland"         "Scotland"                
 [9] "South East"               "South West"              
[11] "Wales"                   
2020-10-06 12:59:55 ERROR Error in .checkTypos(e, names_x): Object 'region' not found. Perhaps you intended region_level_1, ons_region_code

2020-10-06 12:59:55 INFO Processing dataset for united-kingdom-admissions

**Comment by seabbs** [comment]
Ah nice. Looks like it doesn't have region as a variable. when do we usually rename that (checking).

**Comment by joeHickson** [comment]
2020-10-06 13:11:54 TRACE 2
 [1] "East of England"          "England"                 
 [3] "London"                   "Midlands"                
 [5] "North East and Yorkshire" "North West"              
 [7] "Northern Ireland"         "Scotland"                
 [9] "South East"               "South West"              
[11] "Wales"                   
            date           region_level_1 ons_region_code cases_new cases_total
   1: 2020-01-03          East of England              NA         0           0
   2: 2020-01-03                  England              NA         0           0
   3: 2020-01-03                   London              NA         0           0
   4: 2020-01-03                 Midlands              NA         0           0
   5: 2020-01-03 North East and Yorkshire              NA         0           0
  ---                                                                          
3043: 2020-10-05         Northern Ireland              NA         0           0
3044: 2020-10-05                 Scotland              NA         0           0
3045: 2020-10-05               South East              NA         0           0
3046: 2020-10-05               South West              NA         0           0
3047: 2020-10-05                    Wales              NA         0           0
      deaths_new deaths_total recovered_new recovered_total hosp_new hosp_total
   1:          0            0            NA              NA        0          0
   2:          0            0            NA              NA        0          0
   3:          0            0            NA              NA        0          0
   4:          0            0            NA              NA        0          0
   5:          0            0            NA              NA        0          0
  ---                                                                          
3043:          0            0            NA              NA        0          0
3044:          0            0            NA              NA        0          0
3045:          0            0            NA              NA        0          0
3046:          0            0            NA              NA        0          0
3047:          0            0            NA              NA        0          0
      tested_new tested_total hosp_new_first_admissions hosp_new_blend
   1:         NA           NA                        NA             NA
   2:         NA           NA                        NA             NA
   3:         NA           NA                        NA             NA
   4:         NA           NA                        NA             NA
   5:         NA           NA                        NA             NA
  ---                                                                 
3043:         NA           NA                        NA              0
3044:         NA           NA                        NA              0
3045:         NA           NA                        NA             NA
3046:         NA           NA                        NA             NA
3047:         NA           NA                        NA              0
2020-10-06 13:11:54 ERROR Error in .checkTypos(e, names_x): Object 'region' not found. Perhaps you intended region_level_1, ons_region_code

it's odd that it can print the region though...
```R
futile.logger::flog.trace("2")
print(unique(deaths$region))
print(deaths)
national_deaths <- deaths[region %in% c("England", "Scotland", "Wales", "Northern Ireland")]```

**Comment by seabbs** [comment]
is it auto-completing the variable name (is that a thing) in the first cool and not in `data.table` because `data.table` is a bit less mad than base R?

**Comment by joeHickson** [comment]
No idea how! The column rename happens after the modification script is called.

**Comment by seabbs** [comment]
Found it the column gets renamed after this function call: https://github.com/epiforecasts/covid-rt-estimates/blob/8e2d04aa20678c46c2c91f08e59c5fe2a8825d92/R/update-regional.R#L71

**Comment by seabbs** [comment]
might have fixed it?

**Comment by joeHickson** [comment]
we have lift off!
image

**Comment by kathsherratt** [comment]
Thanks @joeHickson and @seabbs ! I've been out the last couple of hours - thanks for debugging and sorting this so quickly.

**Comment by seabbs** [comment]
Any update on this? Any chance we can have something in place for tomorrow morning?

**Comment by joeHickson** [comment]
got one super slow region churning away on the deaths, then it'll start on the hospitalisations.

**Comment by joeHickson** [comment]
Northern Ireland are the problem this time.

**Comment by seabbs** [comment]
NI often shows data issues so this makes sense - @kathsherratt could potentially consider dropping?

**Comment by joeHickson** [comment]
I don't think I can kill it without loosing the summary so I'll just wait a bit. If it's not done by 5 I'll run it again with lower timeouts and check in this evening to push.

**Comment by joeHickson** [comment]
also my guess is it's good to merge and then it'll take first place in the overnight queue

**Comment by seabbs** [comment]
Yes unfortunately that is correct and I am certainly happy for this to be merged now. In terms of priority this is probably in the right place (at the top). Any chance we can collectively deal with #63 asap (i.e tomorrow morning)? Main issue is deciding where things should go and how/when its going to be scheduled.

**Comment by seabbs** [comment]
Damn any informative errors or warnings? I have been aware that the new timeout may not work entirely if a stan object has errored but doesn't have samples (which the default call to `rstan::sampling` would sort out for you). I need to go in and add a check for samples and drop if they are not there before calling `rstan::sflist2stanfit` but no chance yet.

**Comment by joeHickson** [comment]
nothing beyond the sflist issue above. I'm running uk deaths now.

**Comment by seabbs** [comment]
okay - maybe missing some error catching at some level then as that warning should mean that one of the chains failed with an error rather than being timed out.

**Comment by joeHickson** [comment]
all deaths ran quite quickly :)

**Comment by seabbs** [comment]
great - happy to merge? The estimates themselves look plausible (@kathsherratt can you have a look in the branch to check they correspond to what you have been seeing in your repo)?

**Comment by joeHickson** [comment]
there's the technical debt of duplicate functions coming in that you mentioned in your review but I'm willing to accept that

**Comment by kathsherratt** [comment]
Rts look fairly comparable to my previous runs ([here](https://github.com/epiforecasts/rt-comparison-uk-public/tree/master/rt-estimate/estimate-all-time)) though only from a quick first look.

**Comment by seabbs** [comment]
All sounds good merging in 10 mins unless there is any dissent.

**Pull request merged**